crates/ra_ide/src/ssr.rs

```diff
@@ -576,20 +1202,213 @@ mod tests {
+    // the macro call, not from the macro definition.
+    #[test]
+    fn no_match_expression_from_macro() {
+        assert_no_match(
+            "$a.clone()",
+            r#"
+                macro_rules! m1 {
+                    () => {42.clone()}
+                }
+                fn f1() {m1!()}
+                "#,
+        );
+    }
+
+    // We definitely don't want to allow matching of an expression that part originates from the
+    // macro call `42` and part from the macro definition `.clone()`.
+    #[test]
+    fn no_match_split_expression() {
+        assert_no_match(
+            "$a.clone()",
+            r#"
+                macro_rules! m1 {
+                    ($x:expr) => {$x.clone()}
+                }
+                fn f1() {m1!(42)}
+                "#,
+        );
+    }
+
+    #[test]
+    fn replace_function_call() {
         assert_ssr_transform(
-            "foo::<'a>($a, $b)) ==>> foo2($a, $b)",
-            "fn main() { get().bar.foo::<'a>(1); }",
-            "fn main() { foo2(get().bar, 1); }",
-        )
+            "foo() ==>> bar()",
+            "fn f1() {foo(); foo();}",
+            "fn f1() {bar(); bar();}",
+        );
     }
 
     #[test]
-    fn ssr_method_call_and_call() {
+    fn replace_function_call_with_placeholders() {
         assert_ssr_transform(
-            "$o.foo::<i32>($a)) ==>> $o.foo2($a)",
-            "fn main() { X::foo::<i32>(x, 1); }",
-            "fn main() { x.foo2(1); }",
-        )
+            "foo($a, $b) ==>> bar($b, $a)",
+            "fn f1() {foo(5, 42)}",
+            "fn f1() {bar(42, 5)}",
+        );
+    }
+
+    #[test]
+    fn replace_nested_function_calls() {
+        assert_ssr_transform(
+            "foo($a) ==>> bar($a)",
+            "fn f1() {foo(foo(42))}",
+            "fn f1() {bar(bar(42))}",
+        );
+    }
+
+    #[test]
+    fn replace_type() {
+        assert_ssr_transform(
+            "Result<(), $a> ==>> Option<$a>",
+            "fn f1() -> Result<(), Vec<Error>> {foo()}",
+            "fn f1() -> Option<Vec<Error>> {foo()}",
+        );
+    }
+
+    #[test]
+    fn replace_struct_init() {
+        assert_ssr_transform(
+            "Foo {a: $a, b: $b} ==>> Foo::new($a, $b)",
+            "fn f1() {Foo{b: 1, a: 2}}",
+            "fn f1() {Foo::new(2, 1)}",
+        );
+    }
+
+    #[test]
+    fn replace_binary_op() {
+        assert_ssr_transform(
+            "$a + $b ==>> $b + $a",
+            "fn f() {2 * 3 + 4 * 5}",
+            "fn f() {4 * 5 + 2 * 3}",
+        );
+        assert_ssr_transform(
+            "$a + $b ==>> $b + $a",
+            "fn f() {1 + 2 + 3 + 4}",
+            "fn f() {4 + 3 + 2 + 1}",
+        );
+    }
+
+    #[test]
+    fn match_binary_op() {
+        assert_matches(
+            "$a + $b",
+            "fn f() {1 + 2 + 3 + 4}",
+            &["1 + 2 + 3 + 4", "1 + 2 + 3", "1 + 2"],
+        );
     }
 }
```

Comment:
I've split out parsing, matching, replacing and tests. I also moved the code out to a separate crate. The API changed slightly in the process, since I figured I wanted support for multiple rules eventually, so added it now.